src/os/bluestore/BlueStore.cc

```diff
@@ -14662,6 +14697,29 @@ void BlueStore::_wctx_finish(
     }
     delete &lo;
   }
+
+#ifdef HAVE_LIBZBD
+  if (!zones_with_releases.empty()) {
+    // we need to fault the entire extent range in here to determinte if we've dropped
+    // all refs to a zone.
+    o->extent_map.fault_range(db, 0, OBJECT_MAX_SIZE);
+    for (auto& b : o->extent_map.extent_map) {
+      for (auto& e : b.blob->get_blob().get_extents()) {
+	if (e.is_valid()) {
+	  zones_with_releases.erase(e.offset / zone_size);
+	}
+      }
+    }
+    for (auto zone : zones_with_releases) {
+      auto p = o->onode.zone_offset_refs.find(zone);
+      if (p != o->onode.zone_offset_refs.end()) {
+	dout(20) << __func__ << " rm ref zone 0x" << std::hex << zone
+		 << " offset 0x" << p->second << std::dec << dendl;
+	txc->note_release_zone_offset(o, zone, p->second);
```

Comment:
this function call erases from onode's zone_offset_refs[zone] which can be done more efficiently here using the p iterator.

Reply:
additionally as both zones_with_releases and zone_offset_refs are sorted by zone one can use more efficient algorithm to locate the zone in _refs: 
for each zone in zons_with_releases
  scan in _refs from the previous position while current < zone
  if current == zone  
    note_release_zone
    current++
  previous = current